File: fix_defect_e.js
const fs = require('fs');
const path = require('path');

const topics = ['disaster-recovery'];
const base = 'docs/system-design-fundamentals';

const BT = String.fromCharCode(96); // backtick

for (const topic of topics) {
  const dir = path.join(base, topic);
  const files = fs.readdirSync(dir).filter(f => f.endsWith('.mdx'));

  for (const file of files) {
    const fpath = path.join(dir, file);
    let content = fs.readFileSync(fpath, 'utf-8');

    // Fix 1: Opening pattern
    const openRe = new RegExp(
      '(<AsciiDiagram[^>]*?)>\\s*\\n\\s*\\{' + BT,
      'g'
    );
    content = content.replace(openRe, '$1  content={' + BT);

    // Fix 2: Closing pattern — multi-line form
    const closeRe1 = new RegExp(
      BT + '\\s*\\}\\s*\\n\\s*<\\/AsciiDiagram>',
      'g'
    );
    content = content.replace(closeRe1, BT + '} />');

    // Fix 3: Closing pattern — single-line form: `}</AsciiDiagram>
    const closeRe2 = new RegExp(
      BT + '\\s*\\}\\s*<\\/AsciiDiagram>',
      'g'
    );
    content = content.replace(closeRe2, BT + '} />');

    fs.writeFileSync(fpath, content, 'utf-8');
    console.log(`Fixed: ${file} in ${topic}`);
  }
}

console.log('Done!');

File: fix_defect_e.py
#!/usr/bin/env python3
"""Fix Defect E: Convert AsciiDiagram children pattern to content prop pattern."""
import os
import glob
import re

TOPICS = ['serverless', 'deployment-strategies', 'disaster-recovery']
BASE = 'docs/system-design-fundamentals'

for topic in TOPICS:
    path = os.path.join(BASE, topic)
    for f in sorted(glob.glob(os.path.join(path, '*.mdx'))):
        with open(f, 'r', encoding='utf-8') as fh:
            content = fh.read()

        # Pattern 1: Opening: <AsciiDiagram\n  ...props...\n>\n{\n`
        # Replace the >\n{\n` with content={\n`
        # But keep props before it
        content = re.sub(
            r'(<AsciiDiagram[^>]*?)>\s*\n\s*\{\s*\n\s*`',
            r'\1  content={\n`',
            content
        )

        # Pattern 2: Closing: `}\n</AsciiDiagram>
        content = re.sub(
            r'`\s*\}\s*\n\s*</AsciiDiagram>',
            r'`} />',
            content
        )

        # Also handle single-line pattern: prop>{
`
        content = re.sub(
            r'(<AsciiDiagram\s+[^>]*?)>\{\n`',
            r'\1 content={\n`',
            content
        )

        with open(f, 'w', encoding='utf-8') as fh:
            fh.write(content)

        print(f'Fixed: {f}')

print('Done!')
